cpp/inc/bond/core/maybe.h

```diff
@@ -14,143 +18,249 @@ namespace bond
 
 BOND_NORETURN void NothingException();
 
-
+namespace detail
+{
 /// @brief Type used for fields with default value of 'nothing'
 /// See [User's Manual](../../manual/bond_cpp.html#default-value-of-nothing)
 template <typename T>
-class maybe
+class maybe_common
 {
 public:
-    typedef T value_type;
-
-    /// @brief Default constructor
-    maybe()
-        : _nothing(true)
-    {}
+    using value_type = T;
 
-    /// @brief Move constructor
-    maybe(maybe&& that) BOND_NOEXCEPT_IF(std::is_nothrow_move_constructible<T>::value)
-        : _value(std::move(that._value)),
-          _nothing(std::move(that._nothing))
+    /// @brief Check if this object contains nothing
+    bool is_nothing() const
     {
-        that._nothing = true;
+        return static_cast<bool>(_value);
     }
 
-    template <typename Allocator>
-    maybe(maybe&& that, const Allocator& alloc)
-        : _value(std::move(that._value), alloc),
-          _nothing(that._nothing)
+    /// @brief Set to nothing
+    void set_nothing()
     {
-        that._nothing = true;
+        _value = boost::none;
     }
 
-    maybe(const maybe& that) = default;
-    maybe& operator=(const maybe& that) = default;
-
-    /// @brief Constructor from a value T
-    explicit
-    maybe(const T& value)
-        : _value(value),
-          _nothing(false)
+    /// @brief Construct in place
+    ///
+    /// @since 8.0.0
+    template<class... Args>
+    void emplace(Args&&... args)
```

Comment:
Return `T&` (`emplace` should return that)?

Reply:
boost::optional does not return `T&`, so `return _value.value()` it is.

Reply:
Prefer `operator*` or `value()` instead of `get`. The `std::optional` does not have `get`.